### PR TITLE
fix(Core/Reputations): Faction rep gained by killing mobs is now prop…

### LINKED ISSUE #8718: Two Issues With Reputation Gains
### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

### Current Behaviour
Original report: https://github.com/chromiecraft/chromiecraft/issues/2146

1. AC doesn't count reputation if it's not integer number. For example 2.5 count as 2. Players losing this 0.5 every time.
WoWScrnShot_101821_182343
WoWScrnShot_101821_182437
WoWScrnShot_101821_182445

Tester: confirmed: 

WoWScrnShot_102421_152249

I thought this might merely be a display issue, but checking the reputation total directly from the character page also shows it is only incrementing by 2 at a time there as well, so the missing 0.5 is really lost.

2. Kill of NPCs crossing max level of reputation.

creature_onkill_reputation
---
| creature_id | RewOnKillRepFaction1 | MaxStanding1 | RewOnKillRepValue1 |
| ---: | ---: |---: | ---: |
| 3381 | 470 | 5 | 5 | 0 | 0 | 0 | 0 | 

https://user-images.githubusercontent.com/85256266/137775599-ee4593ef-6115-4fbb-9be9-41f5fe70fe2d.mp4

Tester: confirmed. 

Set rep to just below threshhold:
WoWScrnShot_102421_153807

One more kill rewards points beyond cutoff:
WoWScrnShot_102421_153821

### Expected Blizzlike Behaviour

1. AC should count fractional numbers. In my example we need half of 5. In game it should look like this
```
Kill      +2 rep 
Next kill +3 rep
Next kill +2 rep
Next kill +3 rep
```
Missing out on the 0.5 rep per kill is a 20% reduction in rep gain for these factions, which adds up over time.

2. Kill of NPCs can't crossing MaxStanding regardless of amout of reputation. It should stop at `*999`. In my example it should stop at 11999.

### Source

1. Great video from @syssneck [Youtube](https://youtu.be/nGx44JFvd8k?t=83)

Also found a Classic video showing this same behaviour of alternating +2s and +3s: https://youtu.be/GqMSl80KAVU?t=122

2. In my example kill of this Southsea Brigand can't cross revered. [Wiki](https://wowwiki-archive.fandom.com/wiki/Mobs_which_affect_reputation_with_Steamwheedle_Cartel):
Untitled

You can easy find reputation guides or faction descriptions with words:
> Kill of NPC giving to you reputation until friendly/honored/revered

### Steps to reproduce the problem

.go c 13792 will take you to the Southsea pirates outside Ratchet. Kill them and check the other Steamwheedle Cartel rep gains are only +2 per kill.

.modify reputation 470 20998 - takes you to Honoured, within 2 points of Revered wth Ratchet. Now kill another pirate and note you get the full 5 points added to Ratchet rep.

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/e9c71c8ebefc067e3119a2a1a63b74c3239cc55c

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
…erly rounded up.

Do not increase reputation rank if exceeds max rank cap.
Fixes #8718

## Issues Addressed:

- Closes #8718
- Closes https://github.com/chromiecraft/chromiecraft/issues/2146

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

More info [here](https://github.com/azerothcore/azerothcore-wotlk/issues/8718)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by temperrr** [comment]
Merge conflicts

**Comment by UltraNix** [comment]
Done.

**Comment by temperrr** [comment]
Another merge conflict sadly, sorry couldn't test due to holidays
CONFLICT (content): Merge conflict in src/server/game/Entities/Player/Player.cpp
Auto-merging src/server/game/Battlegrounds/Battleground.cpp
Automatic merge failed; fix conflicts and then commit the result.

**Comment by UltraNix** [comment]
@temperrr Done.

**Comment by temperrr** [comment]
- [x] Rep gains alternate between 2 and 3 in the issue linked
- [x] Rep doesnt go higher than what it can go, stops at 11999 in issue linked
- [x] rep still does roll over when it should

**Comment by temperrr** [comment]
I am actually not sure if this pr is correct.
For instance, you would not be able to get revered now with hydraxian waterlords. So I think it should go to the next rank, but dont give any more rep. (so x .../ next rank)

**Comment by UltraNix** [comment]
@temperrr But killing mobs is the only way to get rep with hydraxian waterlords?

**Comment by temperrr** [comment]
Let me test something

**Comment by temperrr** [comment]
Ok if you say its correcct then I am happy :)

File: src/server/game/Reputation/ReputationMgr.cpp [hunk 1]
```diff
@@ -41,6 +41,18 @@ ReputationRank ReputationMgr::ReputationToRank(int32 standing)
...
+{
+    int32 standing = Reputation_Bottom;
+    for (uint8 i = MIN_REPUTATION_RANK; i <= rank; ++i)
+    {
+        standing += PointsInRank[i];
+    }
+
+    return std::max(standing - 1, Reputation_Bottom);
+}
+
```
Comment (acidmanifesto) [inline src/server/game/Reputation/ReputationMgr.cpp:54]: image

Comment (acidmanifesto) [reply src/server/game/Reputation/ReputationMgr.cpp:54]: remove blank line at either line 54 or 55 to pass codestyle check

**Review by acidmanifesto — DISMISSED** [review]

File: src/server/game/Reputation/ReputationMgr.cpp (hunk 1 above)
Comment (Nyeriah) [reply src/server/game/Reputation/ReputationMgr.cpp:54]: @UltraNix

Comment (Nyeriah) [reply src/server/game/Reputation/ReputationMgr.cpp:54]: codecheck broken build cant merge

Comment (FrancescoBorzi) [reply src/server/game/Reputation/ReputationMgr.cpp:54]: @UltraNix please check

> codecheck broken build cant merge

Comment (UltraNix) [reply src/server/game/Reputation/ReputationMgr.cpp:54]: Done.

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**